common/io_lock.c

```diff
@@ -0,0 +1,74 @@
+#include "io_lock.h"
+#include <assert.h>
+#include <ccan/io/io_plan.h>
+
+struct io_lock {
+	bool free;
+};
+
+/* Struct to hold information while we wait for the lock to be freed */
+struct io_lock_waiter {
+	struct io_plan *(*next)(struct io_conn *conn, void *next_arg);
+	void *arg;
+	struct io_lock *lock;
+	enum io_direction dir;
+};
+
+struct io_lock *io_lock_new(const tal_t *ctx)
+{
+	struct io_lock *lock = tal(ctx, struct io_lock);
+	lock->free = true;
+	return lock;
+}
+
+static struct io_plan *io_lock_try_acquire(struct io_conn *conn,
+					   struct io_lock_waiter *waiter)
+{
+	if (waiter->lock->free) {
+		waiter->lock->free = false;
+		return io_out_always(conn, waiter->next, waiter->arg);
```

Comment:
Can we just `return waiter->next(conn, waiter->arg)`?  Hmm, that's a memleak, so we need:

```
    struct io_plan *(*next)(struct io_conn *conn, void *next_arg) = waiter->next;
    void *next_arg = waiter->arg;

    tal_free(waiter);
    return next(conn, next_arg);
}
```